### PR TITLE
Allow switching on and off individual tracers via cli

### PR DESCRIPTION
**Pull request by rvl**
Relates to #1235 

# Overview

- Adds structured logging in more places
- Allow switching on or off individual tracers via cli options.
- Change the type of logging from Trace -> Tracer.

# Comments

```
$ cardano-wallet-jormungandr launch --help-tracing 
Additional tracing options:

  --log-level SEVERITY     Global minimum severity for a message to be logged.
                           Defaults to "DEBUG" unless otherwise configured.
  --trace-NAME=off         Disable logging on the given tracer.
  --trace-NAME=SEVERITY    Set the minimum logging severity for the given
                           tracer. Defaults to "INFO".

The possible log levels (lowest to highest) are:
  debug info notice warning error critical alert emergency

The possible tracers are:
  application  
  api-server  
  api-worker  
  wallet-databases  
  wallet-db  
  network  
  stake-pool-monitor  
  stake-pool-layer  
  stake-pool-db  
  daedalus-ipc  
  worker-registry  
```

### CODE REVIEW

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs [hunk 1]
```diff
@@ -109,36 +123,66 @@ instance ToJSON MsgOut where
...
-        logError trace $ fmt $ "Problem starting Daedalus IPC: "+||err||+""
+        logTrace tr $ MsgStartupError err
   where
     -- Introductory message
     hello = do
-        logDebug trace "Sending Started"
+        logTrace tr MsgSendHello
         pure Started
 
+    msg' m = logTrace tr (MsgReceive m) >> msg m
```
Comment (KtorZ) [inline lib/core/src/Cardano/Wallet/DaedalusIPC.hs]: **[suggestion]** 

```hs
runNodeChannel hello (withLog msg)
{- ... -}
withLog action m = logTrace tr (MsgReceive m) >> action m
```

**Why?** 

Avoid cryptic `msg'` here and convey more meaning in the name of the function at the call site.

File: lib/cli/src/Cardano/CLI.hs [hunk 2]
```diff
@@ -54,10 +56,7 @@ module Cardano.CLI
     , Port (..)
 
     -- * Logging
-    , Verbosity (..)
     , withLogging
-    , verbosityToArgs
-    , verbosityToMinSeverity
```
Comment (KtorZ) [inline lib/cli/src/Cardano/CLI.hs:59]: **[remark]**

These functions should probably be removed completely ?

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs [hunk 3]
```diff
@@ -447,3 +427,69 @@ exitCodeApiServer = \case
...
+
+
+-- | The type of trace events produced by the Jörmungandr API server.
+data Tracers' f = Tracers
+    { applicationTracer :: f ApplicationLog
+    , apiServerTracer :: f ApiLog
+    , stakePoolLayerTracer :: f StakePoolLayerLog
+    , dbLayerTracer :: f DBLog
+    , daedalusIPCTracer :: f DaedalusIPCLog
+    }
```
Comment (KtorZ) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs:436]: **[remark]**

I am not sure to get the choice of names here :thinking: ? 
`dbLayer` refers to both the stake pools database and the wallet database (which we might want to distinguish I suppose). Also, the logs from the network seems missing, oversight or on purpose ?

**Review by KtorZ — COMMENTED** [review]
looks good so far :+1: 
Nice to see these pieces put together.

**Comment by erikd** [comment]
@rvl Please give all PRs a proper title instead of just accepting the default. A proper title makes it easier to judge whether this is something I need to look at or not. Thanks!

**Comment by KtorZ** [comment]
@erikd why would you look at a PR from `cardano-wallet` anyway :thinking: ? 

It's a draft, I assume Rodney just pushed and created the PR before heading off so the team could have an early look.

File: lib/core/src/Cardano/Wallet/DaedalusIPC.hs (hunk 1 above)
Comment (rvl) [reply lib/core/src/Cardano/Wallet/DaedalusIPC.hs]: OK good idea.

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs (hunk 3 above)
Comment (rvl) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr.hs:436]: I revised the list of tracers.

**Review by KtorZ — APPROVED** [review]
LGTM. Played a bit with it, completed missing description for the tracers and grouped some tracers together to make the interface a bit clearer :+1: 

Nice work.

**Pull request merged**